llvm/lib/Target/RISCV/RISCVInstrInfoZb.td

```diff
@@ -554,6 +554,8 @@ def : Pat<(XLenVT (and (shiftop<srl> GPR:$rs1, (XLenVT GPR:$rs2)), 1)),
 
 def : Pat<(XLenVT (shiftop<shl> 1, (XLenVT GPR:$rs2))),
           (BSET (XLenVT X0), GPR:$rs2)>;
+def : Pat<(XLenVT (xor (shiftopw<riscv_sllw> -1, (XLenVT GPR:$rs2)), -1)),
```

Comment:
shiftopw looks through 'and' mask on the lower 5 bits since sllw only uses 5 bits. Bset uses 6 bits on rv64. So this changes behavior.

I think RISCVISD::SLLW also calls SimplifyDemandedBits on the shift amount during DAGCombine which is also an issue.

Reply:
> shiftopw looks through 'and' mask on the lower 5 bits since sllw only uses 5 bits. Bset uses 6 bits on rv64. So this changes behavior.

I'm not sure what you mean, rd, rs1 and rs2 are all 5 bits for bset last time I looked, am I missing something?

Reply:
I'm not referring to the encoding of the instruction. I'm referring to which bits of the register are used. Bset looks at the log2(xlen) bits of rs2 to determine which bit to set.

Reply:
Finally at a computer and not on my phone. My concern is that we will compile this example without an `andi reg, reg, 31` that would be needed with bset but not with sllw. It's really unfortunate that there is no bsetw.

```
define signext i32 @bset_i32_not_mask(i32 signext %a) nounwind {
  %mask = and i32 %a, 31
  %notmask = shl nsw i32 -1, %a
  %sub = xor i32 %notmask, -1
  ret i32 %sub
}
```

https://godbolt.org/z/WcnE577K1